nixos.nix

```diff
@@ -4,454 +4,184 @@ let
+    optionals
     optionalString
     literalExpression
     elem
-    mapAttrs
     intersectLists
     any
     id
-    head
+    ;
+
+  inherit (types)
+    attrsOf
+    submodule
     ;
 
   inherit (utils)
     escapeSystemdPath
-    fsNeededForBoot
+    pathsNeededForBoot
     ;
 
   inherit (pkgs.callPackage ./lib.nix { })
-    splitPath
     concatPaths
     parentsOf
     duplicates
     ;
 
+  inherit (config.users) users;
+
   cfg = config.environment.persistence;
-  users = config.users.users;
-  allPersistentStoragePaths = zipAttrsWith (_name: flatten) (filter (v: v.enable) (attrValues cfg));
+
+  # All persistent storage path submodule values zipped together into
+  # one set. This includes paths from the Home Manager persistence
+  # module and `users` submodules.
+  allPersistentStoragePaths =
+    let
+      # All enabled system paths
+      nixos = filter (v: v.enable) (attrValues cfg);
+
+      # Get the files and directories from the `users` submodules of
+      # enabled system paths
+      nixosUsers = flatten (map attrValues (catAttrs "users" nixos));
+
+      # Fetch enabled paths from all Home Manager users who have the
+      # persistence module loaded
+      homeManager =
+        let
+          paths = flatten
+            (mapAttrsToList
+              (_name: value:
+                attrValues (value.home.persistence or { }))
+              config.home-manager.users or { });
+        in
+        filter (v: v.enable) paths;
+    in
+    zipAttrsWith (_: v: flatten v) (nixos ++ nixosUsers ++ homeManager);
+
   inherit (allPersistentStoragePaths) files directories;
-  mountFile = pkgs.runCommand "impermanence-mount-file" { buildInputs = [ pkgs.bash ]; } ''
+
+  mountFile = pkgs.runCommand "persistence-mount-file" { buildInputs = [ pkgs.bash ]; } ''
     cp ${./mount-file.bash} $out
     patchShebangs $out
   '';
 
+  mkPersistFile = { filePath, persistentStoragePath, method, enableDebugging, ... }:
+    let
+      mountPoint = filePath;
+      targetFile = concatPaths [ persistentStoragePath filePath ];
+      args = escapeShellArgs [
+        mountPoint
+        targetFile
+        method
+        enableDebugging
+      ];
+    in
+    ''
+      ${mountFile} ${args}
+    '';
+
   defaultPerms = {
     mode = "0755";
     user = "root";
     group = "root";
   };
-
-  # Create fileSystems bind mount entry.
-  mkBindMountNameValuePair = { dirPath, persistentStoragePath, hideMount, ... }: {
-    name = concatPaths [ "/" dirPath ];
-    value = {
-      device = concatPaths [ persistentStoragePath dirPath ];
-      noCheck = true;
-      options = [ "bind" "X-fstrim.notrim" ]
-        ++ optional hideMount "x-gvfs-hide";
-      depends = [ persistentStoragePath ];
-    };
-  };
-
-  # Create all fileSystems bind mount entries for a specific
-  # persistent storage path.
-  bindMounts = listToAttrs (map mkBindMountNameValuePair directories);
 in
 {
   options = {
+    home-manager.extraSpecialArgs = mkOption { };
```

Comment:
I assume you're declaring this option because you depend on it existing?

Would it be better to check `options ? home-manager.extraSpecialArgs` instead, wherever you use/read/set this option?

E.g.

```nix
mkMerge [
  (optionalAttrs (options ? home-manager.extraSpecialArgs) {
    home-manager.extraSpecialArgs.persistenceModuleImported = true;
  })
```

Alternatively, if it doesn't need to be a "special" arg, you could instead define `_module.args` in the home-manager submodule; this is usually preferred over using `specialArgs` anyway.

```nix
someHomeManagerModule = {
  _module.args.persistenceModuleImported = true;
}
```

Of course, this won't work if you're using `persistenceModuleImported` to decide which modules to import.

Reply:
> The argument has to be fed to the home manager module from NixOS, not the other way around.

Sure, but module args (`specialArgs` and `_module.args`) are not the only ways to get a value from an outer module configuration into a sub-module configuration.

You can also simply define something in [`home-manager.sharedModules`](https://nix-community.github.io/home-manager/nixos-options.xhtml#nixos-opt-home-manager.sharedModules), e.g. defining some internal option with the value.

From my perspective, using args generally, and specialArgs in particular, should be a last resort.